### ISSUE TITLE
Build on ARM64 Architectures

### ISSUE THREAD

**Issue by abraunegg** [abraunegg/onedrive#95]
Placeholder to track upstream development of ARM64 LDC compiler so that 'onedrive' can be built and tested on ARM64 architectures

LDC Tracker Issues:
* https://github.com/truedat101/dlang-sv-community/issues/45
* https://github.com/ldc-developers/ldc/issues/2153

**Comment by joakim-noah** [comment]
Should work now with LDC 1.11, let me know if you have any problems.

**Comment by abraunegg** [comment]
@joakim-noah 
Looking at the LDC packages, I don't see a specific ARM64 prebuilt package. Can you advise?

**Comment by joakim-noah** [comment]
We haven't put one out, other than the Android Termux package: what OS are you looking to build for?

**Comment by abraunegg** [comment]
Essentially want to build the 'onedrive' client under ARM64 (mainly using Debian as the test platform for that) via automated CI testing using Travis CI. The ARM64 build would occur under qemu, emulating the AMR64 CPU as it does for the armhf builds.

**Comment by joakim-noah** [comment]
I don't see any binary builds in your releases here, do you put some up elsewhere? It'd be easier for you to [cross-compile, if you have a C cross-compiler ready](https://wiki.dlang.org/Building_LDC_runtime_libraries).

**Comment by abraunegg** [comment]
I dont provide binary builds of the application.

If you look through here: https://github.com/abraunegg/onedrive/blob/master/.travis-ci.sh 

You can see how the build is done to test building on various platforms

**Comment by joakim-noah** [comment]
I see, you want to test that your D project will work on AArch64 by downloading a linux/AArch64 build of ldc and then running it in a Qemu AArch64 chroot. We could probably put out a linux/AArch64 build of ldc too, I'll look into it.

**Comment by joakim-noah** [comment]
A Linux/AArch64 build of LDC 1.11 is now up, let me know if you have any problems.

**Comment by abraunegg** [comment]
@joakim-noah 
Thanks - will start working on getting the build scripts up-to-date & working for the ARM64 build

**Comment by abraunegg** [comment]
Travis CI build result using ldc2-1.11.0-linux-aarch64.tar.xz:
```
Running inside chrooted QEMU arm64 environment
Linux travis-job-1498ffcb-8a44-4ca0-a6c9-a545bb95c499 4.4.0-101-generic #124~14.04.1-Ubuntu SMP Fri Nov 10 19:05:36 UTC 2017 aarch64 GNU/Linux
rm -f onedrive onedrive.o onedrive.service onedrive@.service
echo v2.1.2-9-gc68d7be >version
/home/travis/build/abraunegg/onedrive/dlang-arm64/bin/ldmd2 -g -ofonedrive -O -L-lcurl -L-lsqlite3 -L-ldl -J. src/config.d src/itemdb.d src/log.d src/main.d src/monitor.d src/onedrive.d src/qxor.d src/selective.d src/sqlite.d src/sync.d src/upload.d src/util.d src/progress.d
/usr/bin/ld: onedrive.o(.debug_info+0x57): R_AARCH64_ABS64 used with TLS symbol _D4util10deviceNameAya
/usr/bin/ld: onedrive.o(.debug_info+0x81d): R_AARCH64_ABS64 used with TLS symbol _D6upload12fragmentSizel
/usr/bin/ld: onedrive.o(.debug_info+0x83e): R_AARCH64_ABS64 used with TLS symbol _D4sync17thresholdFileSizel
/usr/bin/ld: onedrive.o(.debug_info+0x858): R_AARCH64_ABS64 used with TLS symbol _D4sync14noRemoteDeleteb
/usr/bin/ld: onedrive.o(.debug_info+0x99a): R_AARCH64_ABS64 used with TLS symbol _D3log8usernameAya
/usr/bin/ld: onedrive.o(.debug_info+0x9d4): R_AARCH64_ABS64 used with TLS symbol _D3log11logFilePathAya
/usr/bin/ld: onedrive.o(.debug_info+0x9ee): R_AARCH64_ABS64 used with TLS symbol _D3log7verboseb
/usr/bin/ld: onedrive.o(.debug_info+0xad0): R_AARCH64_ABS64 used with TLS symbol _D3std10functional__T7memoizeS_DQBe5regex__T9regexImplTAyaZQpFNfQjAxaZSQCrQBn8internal2ir__T5RegexTaZQjVii8ZQDjFQCfQBxZ4memoASQEuQEt__TQEkS_DQFjQEf__TQEcTQDvZQEkQDwVii8ZQFsFQEoQEgZ5Value
/usr/bin/ld: onedrive.o(.debug_info+0xed7): R_AARCH64_ABS64 used with TLS symbol _D3std10functional__T7memoizeS_DQBe5regex__T9regexImplTAyaZQpFNfQjAxaZSQCrQBn8internal2ir__T5RegexTaZQjVii8ZQDjFQCfQBxZ11initializedAm
/usr/bin/ld: onedrive.o(.debug_info+0xf11): R_AARCH64_ABS64 used with TLS symbol _D3std5regex8internal2ir__T11wordMatcherZQoFNcNdZ7matcherSQCeQCdQCaQBu11CharMatcher
/usr/bin/ld: onedrive.o(.debug_info+0xf46): R_AARCH64_ABS64 used with TLS symbol _D3std5regex__T9matchOnceTSQzQx8internal2ir__T5RegexTaZQjTAaZQBuFNeQjKxSQCsQCrQBvQBp__TQBpTaZQBvZ8cacheKeySQEbQEa__TQDxTQDqTQCoZQEjFNeQCyKxQCqZ3Key
/usr/bin/ld: onedrive.o(.debug_info+0xfaa): R_AARCH64_ABS64 used with TLS symbol _D3std5regex__T9matchOnceTSQzQx8internal2ir__T5RegexTaZQjTAaZQBuFNeQjKxSQCsQCrQBvQBp__TQBpTaZQBvZ5cacheCQDyQDxQDbQCv__T7MatcherTaZQl
/usr/bin/ld: onedrive.o(.debug_info+0xfea): R_AARCH64_ABS64 used with TLS symbol _D3std5regex8internal2ir__T14defaultFactoryTaZQtFNdNfKxSQCcQCbQByQBs__T5RegexTaZQjZ19backtrackingFactoryCQDzQDyQDvQDp__T14MatcherFactoryTaZQt
/usr/bin/ld: onedrive.o(.debug_info+0x1005): R_AARCH64_ABS64 used with TLS symbol _D3std5regex8internal2ir__T14defaultFactoryTaZQtFNdNfKxSQCcQCbQByQBs__T5RegexTaZQjZ15thompsonFactoryCQDvQDuQDrQDl__T14MatcherFactoryTaZQt
/usr/bin/ld: onedrive.o(.debug_info+0x10dd): R_AARCH64_ABS64 used with TLS symbol _D3std5regex__T9matchOnceTSQzQx8internal2ir__T5RegexTaZQjTAaZQBuFNeQjxSQCrQCqQBuQBo__TQBoTaZQBuZ8cacheKeySQEaQDz__TQDwTQDpTQCnZQEiFNeQCxxQCpZ3Key
/usr/bin/ld: onedrive.o(.debug_info+0x1121): R_AARCH64_ABS64 used with TLS symbol _D3std5regex__T9matchOnceTSQzQx8internal2ir__T5RegexTaZQjTAaZQBuFNeQjxSQCrQCqQBuQBo__TQBoTaZQBuZ5cacheCQDxQDwQDaQCu__T7MatcherTaZQl
/usr/bin/ld: onedrive.o(.debug_info+0x113c): R_AARCH64_ABS64 used with TLS symbol _D3std5regex__T9matchOnceTSQzQx8internal2ir__T5RegexTaZQjTAyaZQBvFNeQkKxSQCtQCsQBwQBq__TQBqTaZQBwZ8cacheKeySQEcQEb__TQDyTQDrTQCpZQEkFNeQCzKxQCqZ3Key
/usr/bin/ld: onedrive.o(.debug_info+0x1180): R_AARCH64_ABS64 used with TLS symbol _D3std5regex__T9matchOnceTSQzQx8internal2ir__T5RegexTaZQjTAyaZQBvFNeQkKxSQCtQCsQBwQBq__TQBqTaZQBwZ5cacheCQDzQDyQDcQCw__T7MatcherTaZQl
sed "s|@PREFIX@|/usr/local|g" systemd.units/onedrive.service.in > onedrive.service
sed "s|@PREFIX@|/usr/local|g" systemd.units/onedrive@.service.in > onedrive@.service
qemu: Unsupported syscall: 278
onedrive v2.1.2-9-gc68d7be
```
Some debug output is generated due to some sort of TLS items, however I think this is a quirk of running under QEMU as an emulated ARM64 processor. Need to validate build on a ARM64 physical system.

Awaiting merge of PR

@joakim-noah 
Is there a test ARM64 physical system which you have access to, which I can test the ARM64 build against?

**Comment by joakim-noah** [comment]
I believe the debug TLS issue is with LLVM, I've seen that on actual hardware too, might be fixed in the upcoming 7.0 release. You can use [the free Shippable CI to test linux/AArch64](http://blog.shippable.com/shippable-arm-packet-deliver-native-ci-cd-for-arm-architecture), [LDC is using it](https://github.com/ldc-developers/ldc/blob/master/shippable.yml). [You can iterate on an Android smartphone first](https://wiki.dlang.org/Build_D_for_Android), if you have a relatively recent one with a 64-bit kernel.
